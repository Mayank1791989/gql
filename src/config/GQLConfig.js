/* @flow */
/* eslint-disable no-use-before-define */
import findConfig from 'find-config';
import JSON5 from 'json5';
import fs from 'fs';
import minimatch from 'minimatch';
import parseGlob from 'parse-glob';
import path from 'path';
import _uniq from 'lodash/uniq';

type Options = {
  cwd?: string
};

const CONFIG_FILE_NAME = '.gqlconfig';

export type QueryParser = (
  'QueryParser'
  | ['EmbeddedQueryParser', { startTag: string, endTag: string }]
);
export type Globs = string | Array<string>;
export type FileMatchConfig = Globs | { include: Globs, ignore?: Globs };
export type ValidateConfig = {
  extends: string,
  rules?: {
    [ruleName: string]: 'off' | 'warn' | 'error',
  },
};
export type QueryConfig = {
  match: FileMatchConfig,
  parser: QueryParser,
  isRelay?: boolean,
  validate?: ValidateConfig,
};

type GQLConfigFile = {
  schema: {
    files: FileMatchConfig,
    validate?: ValidateConfig,
  },
  query?: { // query optional
    files: Array<QueryConfig>,
  },
};

export type { Options as GQLConfigOptions };

class GQLConfig {
  _path: string;
  _configObj: GQLConfigFile;
  _dir: string;

  constructor(options: ?Options) {
    const result = findConfig.obj(CONFIG_FILE_NAME, options);
    if (!result) {
      // throw error .gqlConfig not found
      throw new Error(`Could not find a ${CONFIG_FILE_NAME} file. Create a ${CONFIG_FILE_NAME} file in project root directory.`);
    }

    this._dir = result.dir;
    this._path = result.path;
    this._configObj = this._readConfig(result.path);
  }

  getDir() {
    return this._dir;
  }

  getPath() {
    return this._path;
  }

  getFileExtensions(): Array<string> {
    const { schema, query } = this._configObj;
    const extensions = [];

    // schema
    extensions.push(...extractExtensions(schema.files));

    // query
    if (query) {
      query.files.forEach(({ match }) => {
        extensions.push(...extractExtensions(match));
      });
    }

    return _uniq(extensions);
  }

  getSchema() {
    return this._configObj.schema;
  }

  getQuery() {
    return this._configObj.query;
  }

  match(filePath: string): ?{ type: 'query' | 'schema', opts: Object } {
    const { schema, query } = this._configObj;
    const filePathRelative = path.relative(this.getDir(), filePath);

    // test schema
    if (matcher(filePathRelative, schema.files)) {
      return { type: 'schema', opts: {} };
    }

    // test query
    if (!query) { return null; }

    const { files } = query;
    const opts = files.find(({ match }) => matcher(filePathRelative, match));
    if (opts) {
      return { type: 'query', opts };
    }

    return null;
  }

  _readConfig(filePath: string) { // eslint-disable-line class-methods-use-this
    const fileData = fs.readFileSync(filePath, 'utf8');
    try {
      return (JSON5.parse(fileData): GQLConfigFile); // using flow-runtime it will be validated also
    } catch (err) {
      throw new Error(`\nError parsing ${CONFIG_FILE_NAME}. \n\n${err.message}`);
    }
  }
}

function extractExtensions(match: FileMatchConfig): Array<string> {
  if (typeof match === 'string') {
    return [getExtFromGlob(match)];
  }

  if (Array.isArray(match)) {
    return match.map(getExtFromGlob);
  }

  // object { match, exclude }
  return extractExtensions(match.include);
}

function getExtFromGlob(glob: string): string {
  return parseGlob(glob).path.ext; // without dot
}

function matchGlob(filePath, globs: Globs): boolean {
  if (typeof globs === 'string') {
    return minimatch(filePath, globs);
  }

  // matches any
  return Boolean(globs.find(glob => minimatch(filePath, glob)));
}

function matcher(filePath, match: FileMatchConfig): boolean {
  if (typeof match === 'string' || Array.isArray(match)) {
    return matchGlob(filePath, match);
  }

  const _match = matchGlob(filePath, match.include);
  if (!_match) { return false; }

  const _matchesIgnore = match.ignore ? matchGlob(filePath, match.ignore) : false;

  return _match && !_matchesIgnore;
}

export default GQLConfig;
