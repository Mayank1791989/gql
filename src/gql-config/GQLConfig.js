/* @flow */
import findConfigFile from './findConfigFile';
import readConfigFile from './readConfigFile';
import resolveConfigFile from './resolveConfigFile';
import extractExtensions from './extractExtensions';
import getConfigForFile, {
  type SchemaFileConfig,
  type QueryFileConfig,
} from './getConfigForFile';

import _memoize from 'lodash/memoize';

import { type GQLConfigFileResolved } from './types';
import {
  type ParseOptions as GraphQLParseOptions,
  type BuildSchemaOptions as GraphQLBuildSchemaOptions,
} from 'graphql';

class GQLConfig {
  _dir: string;
  _path: string;
  _configFileResolved: GQLConfigFileResolved;

  constructor(options: { configDir: string }) {
    const result = findConfigFile(options.configDir);
    this._dir = result.dir;
    this._path = result.path;
    this._configFileResolved = resolveConfigFile(
      readConfigFile(result.path),
      result.dir,
    );
    // console.log(JSON.stringify(this._configFileResolved, null, 2));
  }

  // return directory containing config file
  getDir(): string {
    return this._dir;
  }

  // return path of config file
  getPath(): string {
    return this._path;
  }

  // return all file extensions in config file
  getFileExtensions: () => Array<string> = _memoize(() =>
    extractExtensions(this._configFileResolved),
  );

  // return schema config
  getSchemaConfig(): $PropertyType<GQLConfigFileResolved, 'schema'> {
    return this._configFileResolved.schema;
  }

  // return query config
  getQueryConfig(): $PropertyType<GQLConfigFileResolved, 'query'> {
    return this._configFileResolved.query;
  }

  // returns config for given file path
  // file can be schema file or any query file
  getFileConfig(filePath: string): SchemaFileConfig | QueryFileConfig | null {
    return getConfigForFile(filePath, this._configFileResolved, this.getDir());
  }

  getGraphQLParseOptions(): GraphQLParseOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLEmptyFields === 'boolean'
    ) {
      options.allowLegacySDLEmptyFields =
        graphQLOptions.allowLegacySDLEmptyFields;
    }

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLImplementsInterface === 'boolean'
    ) {
      options.allowLegacySDLImplementsInterfaces =
        graphQLOptions.allowLegacySDLImplementsInterface;
    }

    return options;
  }

  getGraphQLBuildSchemaOptions(): GraphQLBuildSchemaOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};
    if (
      graphQLOptions &&
      typeof graphQLOptions.commentDescriptions === 'boolean'
    ) {
      options.commentDescriptions = graphQLOptions.commentDescriptions;
    }
    return options;
  }
}

export default GQLConfig;
