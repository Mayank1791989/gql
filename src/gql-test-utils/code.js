/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
