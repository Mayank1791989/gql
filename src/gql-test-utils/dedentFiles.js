/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
