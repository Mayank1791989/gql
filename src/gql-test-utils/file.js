/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from 'gql-shared/fs-utils';
import { execSync } from 'child_process';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  execSync(`rm -rf ${TEST_TEMP_DIR}`);
}
