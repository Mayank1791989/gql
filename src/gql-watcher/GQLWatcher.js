/* @flow */
import boolValue from 'gql-shared/boolValue';
import log from 'gql-shared/log';
import sane from 'sane';
import os from 'os';
import { execSync } from 'child_process';

import watch, { type WatchOptions } from './watch';
import { memoizeSingle } from 'gql-shared/memoize';
import { type IWatcher } from './types';

const logger = log.getLogger('gql-watcher');

type Options = $ReadOnly<{|
  watchman?: boolean,
  watch?: boolean,
|}>;

export default class GQLWatcher {
  _watchers: Array<*> = [];
  _options: *;

  constructor(options: Options) {
    this._options = options;
  }

  watch(options: $Rest<WatchOptions, {| Watcher: any, watch: any |}>) {
    logger.info(`setting watcher for files ${JSON.stringify(options.files)}`);
    const watcher = watch({
      ...options,
      Watcher: this._getWatcher(boolValue(this._options.watchman, true)),
      watch: boolValue(this._options.watch, true),
    });
    this._watchers.push(watcher);
    return watcher;
  }

  close() {
    this._watchers.forEach(watcher => {
      watcher.close();
    });
  }

  _getWatcher = memoizeSingle((useWatchmanIfAvailable: boolean): Class<
    IWatcher,
  > => {
    if (useWatchmanIfAvailable && this._checkWatchmanInstalled()) {
      logger.info('using "watchman" watcher.');
      return sane.WatchmanWatcher;
    }

    if (os.platform() === 'darwin') {
      logger.info('using "fsevents" watcher.');
      return sane.FSEventWatcher;
    }

    logger.info('using "node" watcher.');
    return sane.NodeWatcher;
  });

  _checkWatchmanInstalled() {
    logger.info('checking "watchman" version...');
    try {
      const version = execSync('watchman --version', {
        stdio: ['ignore'],
        encoding: 'utf8',
      });
      logger.info(`watchman v${version.toString().trim()} found.`);
      return true;
    } catch (e) {
      logger.info('"watchman" not found.');
      return false;
    }
  }
}
