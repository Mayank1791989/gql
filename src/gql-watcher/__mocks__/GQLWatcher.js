/* @flow */
const GQLWatcher = require.requireActual('../GQLWatcher').default;

export default class GQLWatcherMock extends GQLWatcher {
  constructor(options) {
    options.watchman = false;
    super(options);
  }
}
