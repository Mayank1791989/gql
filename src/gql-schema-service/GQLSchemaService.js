/* @flow */
import path from 'path';
import fs from 'fs';

import {
  parse,
  Source,
  GraphQLSchema,
  buildASTSchema as buildASTGraphQLSchema,
  type DocumentNode,
  assertValidSchema,
} from 'graphql';

import { type GQLError, SEVERITY, toGQLError } from 'gql-shared/GQLError';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import {
  type GQLPosition,
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
} from 'gql-shared/types';

import createParser from 'gql-shared/createParser';
import { buildASTSchema, GQLSchema } from 'gql-shared/GQLSchema';
import { newGQLError } from 'gql-shared/GQLError';
import presetExtendSchema from 'gql-shared/presetExtendSchema';

// commands
import findRefsOfTokenAtPosition from './commands/findRefsOfTokenAtPosition';
import getDefinitionAtPosition from './commands/getDefinitionAtPosition';
import getHintsAtPosition from './commands/getHintsAtPosition';
import getInfoOfTokenAtPosition from './commands/getInfoOfTokenAtPosition';

import schemaValidate from './shared/schemaValidate';
import SchemaContext from './shared/SchemaContext';

import GQLConfig from 'gql-config';
import { type SchemaConfigResolved } from 'gql-config/types';
import { type WatchFile } from 'gql-watcher';
import invariant from 'invariant';

type Options = {|
  config: GQLConfig,
  watcher: GQLWatcher,
|};

type CommandParams = {
  fileContent: string,
  fileOptions: SchemaConfigResolved,
  position: GQLPosition,
};

type ParsedSchemaFile = $ReadOnly<{
  error: GQLError | null,
  ast: DocumentNode | null,
  source: Source,
}>;

type ParsedFilesMap = Map<string, ParsedSchemaFile>;

export default class GQLSchemaService extends GQLBaseService {
  _schema: GQLSchema;
  _ast: DocumentNode;
  _errors: Array<GQLError> = [];
  _parsedFilesMap: Map<string, ParsedSchemaFile> = new Map();

  _options: Options;
  _watcher: *;

  constructor(options: Options) {
    super();
    this._options = options;
  }

  async _handleStart() {
    const options = this._options;

    this._watcher = this._options.watcher.watch({
      rootPath: options.config.getDir(),
      files: options.config.getSchemaConfig().files,
      onChange: (files: Array<WatchFile>) => {
        // Handle Error
        try {
          this._updateFiles(files);
        } catch (e) {
          this._emitter.emit('error', e);
        }

        this._emitter.emit('change');
      },
    });

    await this._watcher.onReady();
  }

  async _handleStop() {
    if (this._watcher) {
      await Promise.resolve(this._watcher.close());
    }
  }

  // commands
  findRefsOfTokenAtPosition(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      return findRefsOfTokenAtPosition({
        parser: createParser(params.fileOptions.parser),
        schema: this.getSchema(),
        sourceText: params.fileContent,
        position: params.position,
      });
    }, []);
  }

  getDefinitionAtPosition(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      return getDefinitionAtPosition({
        parser: createParser(params.fileOptions.parser),
        schema: this.getSchema(),
        sourceText: params.fileContent,
        position: params.position,
      });
    }, null);
  }

  getInfoOfTokenAtPosition(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      return getInfoOfTokenAtPosition({
        parser: createParser(params.fileOptions.parser),
        schema: this.getSchema(),
        sourceText: params.fileContent,
        position: params.position,
      });
    }, null);
  }

  getHintsAtPosition(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      return getHintsAtPosition({
        parser: createParser(params.fileOptions.parser),
        schema: this.getSchema(),
        sourceText: params.fileContent,
        position: params.position,
      });
    }, []);
  }

  // utils methods
  getSchema(): GQLSchema {
    return this._schema;
  }

  getGraphQLSchema(): GraphQLSchema {
    const schema = buildASTGraphQLSchema(
      this._ast,
      this._options.config.getGraphQLBuildSchemaOptions(),
    );
    assertValidSchema(schema);
    return schema;
  }

  getSchemaErrors(): Array<GQLError> {
    return this._errors;
  }

  // private methods
  _updateFiles(files: Array<WatchFile>) {
    const options = this._options;

    // console.time('updating files');
    files.forEach(({ name, exists }) => {
      // console.log(name, exists);
      const absPath = path.join(options.config.getDir(), name);
      if (exists) {
        this._parsedFilesMap.set(absPath, this._parseFile(absPath));
      } else {
        this._parsedFilesMap.delete(absPath);
      }
    });
    // console.timeEnd('updating files');

    //  build merged ast
    // console.time('buildAST');
    const { ast, errors: parseErrors } = this._buildASTFromParsedFiles(
      this._parsedFilesMap,
    );
    // console.timeEnd('buildAST');

    // build GQLSchema from ast
    // console.time('buildASTSchema');
    const { schema, errors: buildErrors } = buildASTSchema(
      ast,
      this._options.config.getGraphQLBuildSchemaOptions(),
    );
    // console.timeEnd('buildASTSchema');

    const context = new SchemaContext(schema, options.config.getSchemaConfig());
    // validate
    // console.time('validate');
    // console.log(config.validate);
    const validationErrors = schemaValidate(context, ast);
    // console.timeEnd('validate');

    this._ast = ast;

    this._schema =
      options.config.getSchemaConfig().extendSchema.length > 0
        ? this._presetExtendSchema(schema)
        : schema;

    this._errors = [...parseErrors, ...buildErrors, ...validationErrors];
  }

  _parseFile = (absPath: string) => {
    const content = fs.readFileSync(absPath, 'utf8');
    const source = new Source(content, absPath);
    try {
      const ast = parse(source, this._options.config.getGraphQLParseOptions());
      return {
        ast,
        error: null,
        source,
      };
    } catch (err) {
      return {
        error: toGQLError(err, SEVERITY.error),
        ast: null,
        source,
      };
    }
  };

  _buildASTFromParsedFiles = (parsedFiles: ParsedFilesMap) => {
    const mergedDefinitions = [];
    const errors = [];
    // console.log(parsedFiles.size ==);
    if (parsedFiles.size === 0) {
      const schemaConfig = this._options.config.getSchemaConfig();
      errors.push(
        newGQLError(
          `No schema file found. Make sure schema file glob ${JSON.stringify(
            schemaConfig.files,
          )} matches atleast one file.`,
          null,
          SEVERITY.error,
        ),
      );
    } else {
      for (const parsedFile of parsedFiles.values()) {
        // eslint-disable-line no-restricted-syntax
        const { ast, error } = parsedFile;
        if (error) {
          errors.push(error);
        } else {
          invariant(ast, 'Expecting ast missing');
          const { definitions } = ast;
          mergedDefinitions.push(...definitions);
        }
      }
    }

    return {
      errors, // parsed errors
      ast: {
        kind: 'Document',
        definitions: mergedDefinitions,
      },
    };
  };

  _presetExtendSchema = (schema: GQLSchema) => {
    return this._catchThrownErrors(() => {
      const { config } = this._options;
      return presetExtendSchema(
        schema,
        config.getSchemaConfig().extendSchema,
        config.getGraphQLParseOptions(),
        config.getGraphQLBuildSchemaOptions(),
      );
    }, schema);
  };
}
