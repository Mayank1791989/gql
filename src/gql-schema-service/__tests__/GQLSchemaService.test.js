/* @flow */
import { createTempFiles } from 'gql-test-utils/file';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';

import GQLSchemaService from '../GQLSchemaService';

test('Issues: Crash: Interface', async () => {
  const dir = createTempFiles({
    '.gqlconfig': `
      {
        schema: {
          files: "schema/**/*.gql",
        }
      }
    `,
    'schema/schema.gql': `
      interface Friendly {
        bestFriend: Friendly
      }

      type Person implements Friendly {
        bestFriend: Person
      }
    `,
  });

  const schemaService = new GQLSchemaService({
    config: new GQLConfig({ configDir: dir }),
    watcher: new GQLWatcher({ watch: false }),
  });

  await schemaService.start();
});

describe('Support preset extendSchema', () => {
  const run = async (presetSchema: string) => {
    const dir = createTempFiles({
      'custom-preset': `
        function extendSchema() {
          return ${JSON.stringify(presetSchema, null, 2)};
        }

        module.exports = function customPreset() {
          return {
            extendSchema: extendSchema,
          };
        };
      `,
      '.gqlconfig': `
        {
          schema: {
            files: "schema/**/*.gql",
            presets: ['./custom-preset'],
          }
        }
      `,
      'schema/schema.gql': `
          type Query {
            id: String!
          }
        `,
    });

    const schemaService = new GQLSchemaService({
      config: new GQLConfig({ configDir: dir }),
      watcher: new GQLWatcher({ watch: false }),
    });

    const errors = [];
    schemaService.onError(err => {
      errors.push(err.message);
    });

    await schemaService.start();

    await schemaService.stop();

    return { schemaService, errors };
  };

  test('correctly extend schema if extend Schema valid', async () => {
    const { schemaService, errors } = await run(`
      directive @presetDirective on FIELD
    `);

    expect(errors).toEqual([]);

    expect(
      schemaService.getSchema().getDirective('presetDirective'),
    ).toBeDefined();
  });

  test('report errors if extendSchema failed to parse', async () => {
    const { errors } = await run(`
      directive presetDirective on FIELD
    `);

    expect(errors).toMatchSnapshot();
  });

  test('report errors if errors in extendSchema', async () => {
    const { errors } = await run(`
      extend type Query {
        id: String! # id already present in main schema
      }
    `);

    expect(errors).toMatchSnapshot();
  });
});
