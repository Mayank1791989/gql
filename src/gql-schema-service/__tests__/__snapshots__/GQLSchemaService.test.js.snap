// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Support preset extendSchema report errors if errors in extendSchema 1`] = `
Array [
  "Error in preset extendSchema 

  [4m./custom-preset(3,9)[24m

 [31mâœ–[39m Field \\"Query.id\\" already exists in the schema. It cannot also be defined in this type extension.
[2m   2 |       extend type Query {[22m
[2m > 3 |         id: String! # id already present in main schema[22m
[2m     |         ^[22m
[2m   4 |       }[22m

  [31m1  error[39m
",
]
`;

exports[`Support preset extendSchema report errors if extendSchema failed to parse 1`] = `
Array [
  "Failed to parse preset './custom-preset' extendSchema

 Syntax Error: Expected @, found Name \\"presetDirective\\"

./custom-preset (2:17)
1: 
2:       directive presetDirective on FIELD
                   ^
3:     
",
]
`;
